### DISCUSSION TITLE
Is it possible to replay HTTP requests without connecting to the target server?

### DISCUSSION THREAD

**Discussion by povilasb** [mitmproxy/mitmproxy#5940] [Q&A] [answered]
1. Say I'm running `mitmproxy` in replay mode: `mitmproxy --server-replay requests.txt`

2. When I make a request (`http --proxy https:http://localhost:8080 https://httpbin.org/ip --verify=false`), I can see that `mitmproxy` establishes a TCP connection with the target server:
```log
info: [15:47:14.716][[::1]:51122] server connect httpbin.org:443 (34.205.150.168:443)
info: [15:47:15.940][[::1]:51122] client disconnect
```

3. This is also proved by cutting off from the Internet. Then running the same request results in an error:
```sh
➜  ~ http --proxy https:http://localhost:8080 https://httpbin.org/ip --verify=false

http: LogLevel.ERROR: ProxyError: HTTPSConnectionPool(host='httpbin.org', port=443): Max retries exceeded with url: /ip (Caused by ProxyError('Cannot connect to proxy.', OSError('Tunnel connection failed: 502 Bad Gateway'))) while doing a GET request to URL: https://httpbin.org/ip
```
and `mitmproxy` logs the error too:
```log
info: [15:47:23.048][[::1]:51127] error establishing server connection: [Errno 8] nodename nor
servname provided, or not known
info: [15:47:23.117][[::1]:51127] client disconnect
```

Is there any way to avoid that? Say if I want to use the replay mode for tests, etc. :)

I don't mind tinkering with the **addon**: https://github.com/mitmproxy/mitmproxy/blob/244ff35e606e8ae7ca8aa15d7529566a97ec69f5/mitmproxy/addons/serverplayback.py#L232 - maybe there's a way to say "don't connect to the target server" or so? 🤔

**Comment by povilasb** [comment, marked as answer]
Ah, I missed https://github.com/mitmproxy/mitmproxy/discussions/4992

TL;DR just setting `connection_strategy=lazy` works:
```sh
mitmproxy --set connection_strategy=lazy --server-replay requests.txt
```
